syntax = "proto3";

package spaceart;

service Presenter {
  rpc CheckIn(Account) returns (CheckInResponse) {}
  rpc SubmitAccount(Account) returns (CheckInResponse) {}
  rpc Present(PresentationToken) returns (stream PresentationLine) {}
}

service Registrar {
  rpc Validate(Account) returns (AuthResponse) {}
  rpc Register(Account) returns (AuthResponse) {}
}

service Gallery {
  rpc GetArt(ArtRequest) returns (stream ArtLine) {}
}

message Account {
  string username = 1;
  optional string password = 2;
}

message PresentationToken {
  string token = 1;
}

message PresentationLine {
  string line = 1;
}

message CheckInResponse {
  bool valid = 1;
  optional bool wrong_password = 2;
  optional bool new_password_desired = 3;
  optional bool user_exists = 4;
  optional string token = 5;
}

message AuthResponse {
  bool valid = 1;
  optional bool wrong_password = 2;
  optional bool user_exists = 3;
}

message ArtRequest {
  string name = 1;
}

message ArtLine {
  string line = 1;
}
